### PR TITLE
Enable require_trailing_commas lint rule

### PR DESCRIPTION
**Pull request by chenzhiguang**
Add this optional trailing commas to parameter lists to get good formatting. It is also recommended in [Flutter: Using trailing commas
](https://docs.flutter.dev/development/tools/formatting#using-trailing-commas)

### CODE REVIEW

**Comment by kevmoo** [comment]
I like this lint, but since @srawlins is the primary maintainer here, I'll let him comment!

File: benchmark/benchmark.dart [hunk 1]
```diff
@@ -56,8 +56,10 @@ String _loadFile(String name) {
 }
 
 void _printResult(String label, double time) {
-  print('$label: ${_padLeft(time.toStringAsFixed(2), 4)}ms '
-      "${'=' * ((time * 20).toInt())}");
+  print(
+    '$label: ${_padLeft(time.toStringAsFixed(2), 4)}ms '
+    "${'=' * ((time * 20).toInt())}",
```
Comment (srawlins) [inline benchmark/benchmark.dart:61]: I am not a fan of this lint rule. For cases like this.

I am a fan of adding trailing commas to argument lists when there is more than one argument, and they span more than 1 line without the comma.

This change here is harder for me to read though.

File: bin/markdown.dart [hunk 2]
```diff
@@ -19,16 +19,18 @@ Future<void> main(List<String> args) async {
...
-        allowed: ['none', 'CommonMark', 'GitHubFlavored', 'GitHubWeb'],
-        defaultsTo: 'CommonMark',
-        help: 'Specify a set of extensions',
-        allowedHelp: {
-          'none': 'No extensions; similar to Markdown.pl',
-          'CommonMark': 'Parse like CommonMark Markdown (default)',
-          'GitHubFlavored': 'Parse like GitHub Flavored Markdown',
-          'GitHubWeb': 'Parse like GitHub\'s Markdown-enabled web input fields',
-        });
+    ..addOption(
```
Comment (srawlins) [inline bin/markdown.dart:22]: I like this change though.

File: bin/markdown.dart [hunk 3]
```diff
@@ -75,5 +78,6 @@ By default, CommonMark Markdown will be parsed. This can be changed with
 the --extensionSet flag.
 
 ${parser.usage}
-''');
+''',
+  );
```
Comment (srawlins) [inline bin/markdown.dart:82]: Again, I don't think this is helpful.

**Review by srawlins — COMMENTED** [review]
I like many of the changes that this induces, but I strongly dislike some of them. What do you say to keeping many of these changes (I can mark the ones to revert), and we do not change analysis_options.yaml?

**Comment by chenzhiguang** [comment]
> I like many of the changes that this induces, but I strongly dislike some of them. What do you say to keeping many of these changes (I can mark the ones to revert), and we do not change analysis_options.yaml?

I do not like some of them too. 

Yep. let me remove the rule from analysis_options.yaml, and reformat it.

File: lib/src/block_parser.dart [hunk 4]
```diff
@@ -53,7 +53,8 @@ final _olPattern =
 
 /// A line of hyphens separated by at least one pipe.
 final _tablePattern = RegExp(
-    r'^[ ]{0,3}\|?([ \t]*:?\-+:?[ \t]*\|)+([ \t]|[ \t]*:?\-+:?[ \t]*)?$');
+  r'^[ ]{0,3}\|?([ \t]*:?\-+:?[ \t]*\|)+([ \t]|[ \t]*:?\-+:?[ \t]*)?$',
```
Comment (srawlins) [inline lib/src/block_parser.dart:56]: I think this should be reverted.

File: lib/src/inline_parser.dart [hunk 5]
```diff
@@ -1103,10 +1152,12 @@ class LinkSyntax extends DelimiterSyntax {
       // Normally, label text does not get parsed as inline Markdown. However,
       // for the benefit of the link resolver, we need to at least escape
       // brackets, so that, e.g. a link resolver can receive `[\[\]]` as `[]`.
-      final resolved = linkResolver(label
-          .replaceAll(r'\\', r'\')
-          .replaceAll(r'\[', '[')
-          .replaceAll(r'\]', ']'));
+      final resolved = linkResolver(
```
Comment (srawlins) [inline lib/src/inline_parser.dart:1155]: I think this should be reverted. It just feels very strange to me in cases like this where there is only one parameter...

File: tool/stats_lib.dart [hunk 6]
```diff
@@ -15,18 +15,20 @@ import '../test/util.dart';
 
 // Locate the "tool" directory. Use mirrors so that this works with the test
 // package, which loads this suite into an isolate.
-String get toolDir =>
-    p.dirname((reflect(loadCommonMarkSections) as ClosureMirror)
-        .function
-        .location!
-        .sourceUri
-        .path);
+String get toolDir => p.dirname(
```
Comment (srawlins) [inline tool/stats_lib.dart:18]: I think this should be reverted.

But this change I think indicates a smell; this arrow-function should probably be a function body with a local variable, like a variable for the `reflect` result...

File: tool/update_emojis.dart [hunk 7]
```diff
@@ -14,18 +14,25 @@ Future<void> main() async {
...
-          await response.cast<List<int>>().transform(utf8.decoder).join(''))
-      .map((String alias, dynamic info) =>
-          MapEntry(alias, info.cast<String, dynamic>()))
+    await response.cast<List<int>>().transform(utf8.decoder).join(''),
+  )
+      .map(
+        (String alias, dynamic info) => MapEntry(
+          alias,
+          info.cast<String, dynamic>(),
+        ),
```
Comment (srawlins) [inline tool/update_emojis.dart:23]: I think this should be reverted.

File: tool/update_emojis.dart [hunk 8]
```diff
@@ -14,18 +14,25 @@ Future<void> main() async {
...
+  final emojisContent = StringBuffer(
+    '''
 // GENERATED FILE. DO NOT EDIT.
 //
 // This file was generated from emojilib's emoji data file:
 // $_emojisJsonRawUrl
 // at ${DateTime.now()} by the script, tool/update_emojis.dart.
 
-''');
+''',
```
Comment (srawlins) [inline tool/update_emojis.dart:34]: I think this should be reverted.

File: tool/update_emojis.dart [hunk 9]
```diff
@@ -39,7 +46,9 @@ Future<void> main() async {
   });
   emojisContent.writeln('};');
   File(_emojisFilePath).writeAsStringSync(emojisContent.toString());
-  print('Wrote data to $_emojisFilePath for $emojiCount emojis, '
-      'ignoring ${ignored.length}: ${ignored.join(', ')}.');
+  print(
+    'Wrote data to $_emojisFilePath for $emojiCount emojis, '
+    'ignoring ${ignored.length}: ${ignored.join(', ')}.',
```
Comment (srawlins) [inline tool/update_emojis.dart:51]: I think this should be reverted.

**Review by srawlins — CHANGES_REQUESTED** [review]
The vast majority of the changes improve readability I think.

Just a few where I think the change is much worse.

**Comment by chenzhiguang** [comment]
How about close this PR?  I have just created a new [PR](https://github.com/dart-lang/markdown/pull/419) which has much fewer changes. please have a look.

**Comment by srawlins** [comment]
Closing in favor of #420

**Pull request closed without merging**